src/cloudai/cli/handlers.py

```
@@ -118,6 +118,40 @@ def prepare_installation(
+    HAS_CUSTOM_TRAINING_LOOP: bool
+
+    def train(self) -> None: ...
+
+
+def _has_custom_training_loop(agent: object) -> bool:
+    return bool(getattr(agent, "HAS_CUSTOM_TRAINING_LOOP", False))
+
+
+def _run_custom_training_loop(agent: CustomTrainingLoopAgent, agent_type: str) -> int:
+    """Drive an agent's self-contained training loop and return a process-style exit code."""
+    logging.info(f"Agent {agent_type} drives its own training loop; delegating to agent.train().")
```

Ruff (0.15.12)

[warning] Logging statement uses f-string

(G004)